server/api-manager.js

```diff
@@ -917,13 +923,39 @@ class MindbodyQueries {
   }
 
 
-
-  //@TODO implement caching based on hashes of url's here
   decorateAndMake(request) {
+    //instead of making a new GET request, used cached results when possible
+    if (request.verb === "GET") {
+      let hashCode = this.getStringHash(request.siteId.toString + request.url);
+      let cacheTuple = this.cachedResults[hashCode];
+      //console.log("MindbodyAccess cache for " + request.url + " (" + hashCode + ") held value: " + JSON.stringify(cacheTuple));
+      if (cacheTuple && cacheTuple.siteId === request.siteId && cacheTuple.url === request.url) {
+        const cacheDuration = 300000; //5 minutes, expressed in milliseconds
+        if (Date.now() - cacheTuple.creationTime < cacheDuration) {
+          //console.log("MindbodyAccess cache hit for " + request.url);
+          return Promise.resolve(cacheTuple.result);
+        }
+      }
+    }
+
     request.addAuth(this.authToken);
     this.requestNum++;
     if (!this.atLimit()) {
-      return request.makeRequest();
+      let reqPromise = request.makeRequest();
```

Comment:
I still struggle to use JS Promises "the right way" This code behaves correctly but it may be doing something dumb with promises. 
This code does a slightly odd Promise thing that's not the normal chaining. Here, we make a get request, we attach a "then" clause to store the results when they arrive, and we return the promise from the original get request. I think that the cache's then clause will run before calling code gets access to the results.
Anyway, let me know if this looks wrong to you.